src/System.Private.CoreLib/src/System/Threading/ManagedThreadId.cs

```diff
@@ -9,15 +9,173 @@
+
+            private readonly int _used; // Number of ids tracked by this node and all its childs
+            private readonly int _size; // Maximum number of ids that can be tracked by this node and all its childs
+
+            private readonly uint _bitmap; // Bitmap of ids tracked by this node
+
+            private const int BitsPerNode = 32;
+
+            private ImmutableIdDispenser(ImmutableIdDispenser left, ImmutableIdDispenser right, int used, int size, uint bitmap)
+            {
+                _left = left;
+                _right = right;
+                _used = used;
+                _size = size;
+                _bitmap = bitmap;
+
+                CheckInvariants();
+            }
+
+            [Conditional("DEBUG")]
+            private void CheckInvariants()
+            {
+                int actualUsed = 0;
+
+                uint countBits = _bitmap;
+                while (countBits != 0)
+                {
+                    actualUsed += (int)(countBits & 1);
+                    countBits >>= 1;
+                }
+
+                if (_left != null)
+                {
+                    Debug.Assert(_left._size == ChildSize);
+                    actualUsed += _left._used;
+                }
+                if (_right != null)
+                {
+                    Debug.Assert(_right._size == ChildSize);
+                    actualUsed += _right._used;
+                }
+
+                Debug.Assert(actualUsed == _used);
+                Debug.Assert(_used <= _size);
+            }
+
+            private int ChildSize
+            {
+                get
+                {
+                    Debug.Assert((_size / 2) >= (BitsPerNode / 2));
+                    return (_size / 2) - (BitsPerNode / 2);
+                }
+            }
+
+            public static ImmutableIdDispenser Empty
+            {
+                get
+                {
+                    // The empty dispenser has the id=0 allocated, so it is not really empty.
+                    // It saves us from dealing with the corner case of true empty dispenser,
+                    // and it ensures that ManagedThreadIdNone will not be ever given out.
+                    return new ImmutableIdDispenser(null, null, 1, BitsPerNode, 1);
+                }
+            }
+
+            public ImmutableIdDispenser AllocateId(out int id)
+            {
+                if (_used == _size)
+                {
+                    id = _size;
+                    return new ImmutableIdDispenser(this, null, _size + 1, checked(2 * _size + BitsPerNode), 1);
+                }
+
+                var bitmap = _bitmap;
+                var left = _left;
+                var right = _right;
+
+                // Any free bits in current node?
+                if (bitmap != UInt32.MaxValue)
+                {
+                    int bit = 0;
+                    while ((bitmap & (uint)(1 << bit)) != 0)
```

Comment:
I got that one. You are covering my case with the Empty which makes it fully used so always a new child node will be allocated. LGTM.